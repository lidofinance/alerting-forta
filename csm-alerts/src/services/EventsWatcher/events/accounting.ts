import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSAccounting__factory } from '../../../generated/typechain'
import * as CSAccounting from '../../../generated/typechain/CSAccounting'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress, formatEther } from '../../../utils/string'

const ICSAccounting = CSAccounting__factory.createInterface()

export function getCSAccountingEvents(accounting: string): EventOfNotice[] {
    return [
        {
            address: accounting,
            abi: ICSAccounting.getEvent('ChargePenaltyRecipientSet').format('full'),
            alertId: 'CS-ACCOUNTING-CHARGE-PENALTY-RECIPIENT-SET',
            name: 'ðŸš¨ CSAccounting: Charge penalty recipient set',
            description: (args: CSAccounting.ChargePenaltyRecipientSetEvent.OutputObject) =>
                `Charge penalty recipient set to ${etherscanAddress(args.chargePenaltyRecipient)} (expecting the treasury)`,
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveUpdated').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-UPDATED',
            name: 'ðŸš¨ CSAccounting: Bond curve updated',
            description: (args: CSAccounting.BondCurveUpdatedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve #${args.curveId} updated. New values: [${bondCurveString}]`
            },
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveAdded').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-ADDED',
            name: 'ðŸ”´ CSAccounting: Bond curve added',
            description: (args: CSAccounting.BondCurveAddedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve added: [${bondCurveString}]`
            },
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
    ]
}
