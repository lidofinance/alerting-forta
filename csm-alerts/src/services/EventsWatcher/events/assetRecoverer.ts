import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { AssetRecoverer__factory } from '../../../generated/typechain'
import * as AssetRecoverer from '../../../generated/typechain/AssetRecoverer'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress, formatEther, formatShares } from '../../../utils/string'

const IAssetRecoverer = AssetRecoverer__factory.createInterface()

export function getAssetRecovererEvents(
    contracts: { name: string; address: string }[],
): EventOfNotice[] {
    return contracts.flatMap((contract) => {
        return [
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC20Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC20-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC20 recovered',
                description: (args: AssetRecoverer.ERC20RecoveredEvent.OutputObject) =>
                    `ERC20 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Amount: ${args.amount}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC721Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC721-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC721 recovered',
                description: (args: AssetRecoverer.ERC721RecoveredEvent.OutputObject) =>
                    `ERC721 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Token ID: ${args.tokenId}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('ERC1155Recovered').format('full'),
                alertId: 'ASSET-RECOVERER-ERC1155-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: ERC1155 recovered',
                description: (args: AssetRecoverer.ERC1155RecoveredEvent.OutputObject) =>
                    `ERC1155 recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Token: ${etherscanAddress(args.token)}\n` +
                    `Token ID: ${args.tokenId}\n` +
                    `Amount: ${args.amount}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('EtherRecovered').format('full'),
                alertId: 'ASSET-RECOVERER-ETHER-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: Ether recovered',
                description: (args: AssetRecoverer.EtherRecoveredEvent.OutputObject) =>
                    `Ether recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Amount: ${formatEther(args.amount)}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
            {
                address: contract.address,
                abi: IAssetRecoverer.getEvent('StETHSharesRecovered').format('full'),
                alertId: 'ASSET-RECOVERER-STETH-SHARES-RECOVERED',
                name: 'ðŸ”´ AssetRecoverer: stETH Shares recovered',
                description: (args: AssetRecoverer.StETHSharesRecoveredEvent.OutputObject) =>
                    `StETH Shares recovered on ${contract.name}:\n` +
                    `Recipient: ${etherscanAddress(args.recipient)}\n` +
                    `Amount: ${formatShares(args.shares)}`,
                severity: FindingSeverity.High,
                type: FindingType.Info,
            },
        ]
    })
}
