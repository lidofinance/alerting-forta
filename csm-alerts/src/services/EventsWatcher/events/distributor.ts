import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSFeeDistributor__factory } from '../../../generated/typechain'
import * as CSFeeDistributor from '../../../generated/typechain/CSFeeDistributor'
import { EventOfNotice } from '../../../shared/types'
import { ipfsLink } from '../../../utils/string'

const ICSFeeDistributor = CSFeeDistributor__factory.createInterface()

export function getCSFeeDistributorEvents(distributorAddress: string): EventOfNotice[] {
    return [
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionDataUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-DATA-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution data updated',
            description: (args: CSFeeDistributor.DistributionDataUpdatedEvent.OutputObject) =>
                `Total Claimable Shares: ${args.totalClaimableShares}\n` +
                `Tree Root: ${args.treeRoot}\n` +
                `Tree CID: ${ipfsLink(args.treeCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionLogUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-LOG-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution log updated',
            description: (args: CSFeeDistributor.DistributionLogUpdatedEvent.OutputObject) =>
                `Log CID: ${ipfsLink(args.logCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
    ]
}
