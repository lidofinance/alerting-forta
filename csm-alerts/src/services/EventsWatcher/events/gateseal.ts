import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { GateSeal__factory } from '../../../generated/typechain'
import * as GateSeal from '../../../generated/typechain/GateSeal'
import { EventOfNotice } from '../../../shared/types'
import { addressOnExplorer } from '../../../utils/string'
import { formatDelay } from '../../../utils/time'

const IGateSeal = GateSeal__factory.createInterface()

export function getGateSealEvents(
    address: string,
    knownContracts: { [key: string]: string },
): EventOfNotice[] {
    return [
        {
            address: address,
            abi: IGateSeal.getEvent('Sealed').format('full'),
            alertId: 'CS-GATE-SEAL-SEALED',
            name: 'ðŸš¨ CSM Gate Seal paused a contract',
            description: (args: GateSeal.SealedEvent.OutputObject) =>
                `${addressOnExplorer(args.sealable)} (${knownContracts[args.sealable] ?? 'unknown'})` +
                `was paused for ${formatDelay(args.sealed_for)} by ${addressOnExplorer(args.sealed_by)}`,
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
    ]
}
