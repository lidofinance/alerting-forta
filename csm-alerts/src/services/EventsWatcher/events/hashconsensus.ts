import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { HashConsensus__factory } from '../../../generated/typechain'
import * as HashConsensus from '../../../generated/typechain/HashConsensus'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress } from '../../../utils/string'

const IHashConsensus = HashConsensus__factory.createInterface()

export function getHashConsensusEvents(
    address: string,
    knownMembers: { [key: string]: string },
): EventOfNotice[] {
    return [
        {
            address,
            abi: IHashConsensus.getEvent('MemberAdded').format('full'),
            alertId: 'HASH-CONSENSUS-MEMBER-ADDED',
            name: 'ðŸ”´ CSM HashConsensus: Member added',
            description: (args: HashConsensus.MemberAddedEvent.OutputObject) =>
                `New member ${etherscanAddress(args.addr)} (${knownMembers[args.addr] ?? 'unknown'}) added\n` +
                `Total members: ${args.newTotalMembers}\n` +
                `New quorum: ${args.newQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('MemberRemoved').format('full'),
            alertId: 'HASH-CONSENSUS-MEMBER-REMOVED',
            name: 'ðŸ”´ CSM HashConsensus: Member removed',
            description: (args: HashConsensus.MemberRemovedEvent.OutputObject) =>
                `Member ${etherscanAddress(args.addr)} (${knownMembers[args.addr] ?? 'unknown'}) removed\n` +
                `Total members: ${args.newTotalMembers}\n` +
                `New quorum: ${args.newQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('QuorumSet').format('full'),
            alertId: 'HASH-CONSENSUS-QUORUM-SET',
            name: 'ðŸ”´ CSM HashConsensus: Quorum set',
            description: (args: HashConsensus.QuorumSetEvent.OutputObject) =>
                `Quorum set to ${args.newQuorum}.\n` +
                `Total members: ${args.totalMembers}\n` +
                `Previous quorum: ${args.prevQuorum}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('FastLaneConfigSet').format('full'),
            alertId: 'HASH-CONSENSUS-FASTLANE-CONFIG-SET',
            name: 'ðŸ”´ CSM HashConsensus: Fastlane config set',
            description: (args: HashConsensus.FastLaneConfigSetEvent.OutputObject) =>
                `Fastlane configuration set with length slots: ${args.fastLaneLengthSlots}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('FrameConfigSet').format('full'),
            alertId: 'HASH-CONSENSUS-FRAME-CONFIG-SET',
            name: 'ðŸ”´ CSM HashConsensus: Frame config set',
            description: (args: HashConsensus.FrameConfigSetEvent.OutputObject) =>
                `Frame configuration set:\n` +
                `New initial epoch: ${args.newInitialEpoch}\n` +
                `Epochs per frame: ${args.newEpochsPerFrame}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ReportProcessorSet').format('full'),
            alertId: 'HASH-CONSENSUS-REPORT-PROCESSOR-SET',
            name: 'ðŸ”´ CSM HashConsensus: Report processor set',
            description: (args: HashConsensus.ReportProcessorSetEvent.OutputObject) =>
                `Previous processor: ${etherscanAddress(args.prevProcessor)}\n` +
                `Current processor: ${etherscanAddress(args.processor)}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ConsensusLost').format('full'),
            alertId: 'HASH-CONSENSUS-LOST',
            name: 'ðŸ”´ CSM HashConsensus: Consensus lost',
            description: (args: HashConsensus.ConsensusLostEvent.OutputObject) =>
                `Consensus lost for slot ${args.refSlot}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
        {
            address,
            abi: IHashConsensus.getEvent('ConsensusReached').format('full'),
            alertId: 'HASH-CONSENSUS-REACHED',
            name: 'ðŸ”µ CSM HashConsensus: Consensus reached',
            description: (args: HashConsensus.ConsensusReachedEvent.OutputObject) =>
                `Consensus reached for slot ${args.refSlot}\n` +
                `Report hash: ${args.report}\n` +
                `Support: ${args.support}\n` +
                `Contract: ${etherscanAddress(address)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
    ]
}
