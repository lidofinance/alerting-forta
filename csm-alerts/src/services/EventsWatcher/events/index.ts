export * from './assetRecoverer'
export * from './distributor'
export * from './accounting'
export * from './pausable'
export * from './proxies'
export * from './module'
export * from './oracle'
export * from './roles'
