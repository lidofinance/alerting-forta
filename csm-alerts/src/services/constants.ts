import { DeployedAddresses } from '../shared/types'

export const DEPLOYED_ADDRESSES: DeployedAddresses = {
    CS_MODULE: '0xdA7dE2ECdDfccC6c3AF10108Db212ACBBf9EA83F',
    CS_ACCOUNTING: '0x4d72BFF1BeaC69925F8Bd12526a39BAAb069e5Da',
    CS_FEE_DISTRIBUTOR: '0xD99CC66fEC647E68294C6477B40fC7E0F6F618D0',
    CS_FEE_ORACLE: '0x4D4074628678Bd302921c20573EEa1ed38DdF7FB',
    CS_VERIFIER: '0x3Dfc50f22aCA652a0a6F28a0F892ab62074b5583',
    CS_EARLY_ADOPTION: '0x3D5148ad93e2ae5DedD1f7A8B3C19E7F67F90c0E',
    CS_GATE_SEAL: '0x16Dbd4B85a448bE564f1742d5c8cCdD2bB3185D0',
    LIDO_STETH: '0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84',
    BURNER: '0xD15a672319Cf0352560eE76d9e89eAB0889046D3',
    HASH_CONSENSUS: '0x71093efF8D8599b5fA340D665Ad60fA7C80688e4',
    STAKING_ROUTER: '0xFdDf38947aFB03C621C71b06C9C70bce73f12999',
    MULTICALL3: '0xcA11bde05977b3631167028862bE2a173976CA11',
    PROVER: '0xDCBDb89D30E921C73B496e9d20a61d283A5D2401',
}

export const ALIASES: Record<keyof DeployedAddresses, string> = {
    CS_MODULE: 'CSM',
    CS_ACCOUNTING: 'CSAccounting',
    CS_FEE_DISTRIBUTOR: 'CSFeeDistributor',
    CS_FEE_ORACLE: 'CSFeeOracle',
    CS_VERIFIER: 'CSAccounting',
    CS_EARLY_ADOPTION: 'CSEarlyAdoption',
    CS_GATE_SEAL: 'CSM Gate Seal',
    LIDO_STETH: 'Lido',
    BURNER: 'Burner',
    HASH_CONSENSUS: 'CSM HashConsensus',
    STAKING_ROUTER: 'StakingRouter',
    MULTICALL3: 'Multicall3',
    PROVER: 'Prover Wallet',
}

export const ORACLE_MEMBERS: { [key: string]: string } = {
    '0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E': 'Chorus One',
    '0x404335BcE530400a5814375E7Ec1FB55fAff3eA2': 'Staking Facilities',
    '0x946D3b081ed19173dC83Cd974fC69e1e760B7d78': 'Stakefish',
    '0x007DE4a5F7bc37E2F26c0cb2E8A95006EE9B89b5': 'P2P',
    '0xc79F702202E3A6B0B6310B537E786B9ACAA19BAf': 'Chainlayer',
    '0x61c91ECd902EB56e314bB2D5c5C07785444Ea1c8': 'bloXroute',
    '0x1Ca0fEC59b86F549e1F1184d97cb47794C8Af58d': 'Instadapp', // Revoked one.
    '0x73181107c8D9ED4ce0bbeF7A0b4ccf3320C41d12': 'Instadapp',
    '0xe57B3792aDCc5da47EF4fF588883F0ee0c9835C9': 'MatrixedLink',
    '0xA7410857ABbf75043d61ea54e07D57A6EB6EF186': 'Kyber Network',
}
