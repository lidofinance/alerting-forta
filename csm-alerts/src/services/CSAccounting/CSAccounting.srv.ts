import { elapsedTime } from '../../utils/time'
import { either as E } from 'fp-ts'
import { Logger } from 'winston'
import { BlockDto, EventOfNotice, TransactionDto, handleEventsOfNotice } from '../../entity/events'
import { Finding } from '../../generated/proto/alert_pb'
import { filterLog } from 'forta-agent'
import { APPROVAL_EVENT } from '../../utils/events/cs_accounting_events'

export abstract class ICSAccountingClient {
  public abstract getBlockByNumber(blockNumber: number): Promise<E.Either<Error, BlockDto>>
}

export class CSAccountingSrv {
  private readonly name = 'CSAccountingSrv'
  private readonly logger: Logger
  private readonly csAccountingClient: ICSAccountingClient

  private readonly csAccountingEvents: EventOfNotice[]
  private readonly csAccountingAddress: string
  private readonly stETHAddress: string

  constructor(
    logger: Logger,
    ethProvider: ICSAccountingClient,
    csAccountingEvents: EventOfNotice[],
    csAccountingAddress: string,
    stETHAddress: string,
  ) {
    this.logger = logger
    this.csAccountingClient = ethProvider

    this.csAccountingEvents = csAccountingEvents
    this.csAccountingAddress = csAccountingAddress
    this.stETHAddress = stETHAddress
  }

  public async initialize(currentBlock: number): Promise<Finding[] | Error | null> {
    const start = new Date().getTime()

    const currBlock = await this.csAccountingClient.getBlockByNumber(currentBlock)
    if (E.isLeft(currBlock)) {
      this.logger.error(elapsedTime(`Failed [${this.name}.initialize]`, start))
      return currBlock.left
    }

    this.logger.info(elapsedTime(`[${this.name}.initialize]`, start))
    return null
  }

  public getName(): string {
    return this.name
  }

  async handleBlock(blockDto: BlockDto): Promise<Finding[]> {
    const start = new Date().getTime()
    const findings: Finding[] = []

    const [rolesChangingFindings] = await Promise.all([this.handleRolesChanging(blockDto.number)])

    findings.push(...rolesChangingFindings)

    this.logger.info(elapsedTime(CSAccountingSrv.name + '.' + this.handleBlock.name, start))

    return findings
  }
  // to be implemented
  handleRolesChanging(blockNumber: number): Promise<Finding[]> {
    const out: Finding = new Finding()
    this.logger.info(`${blockNumber}`)
    return Promise.resolve([out])
  }

  public handleStETHApprovalEvents(txEvent: TransactionDto): Finding[] {
    const out: Finding[] = []

    const approvalEvents = filterLog(txEvent.logs, APPROVAL_EVENT, this.stETHAddress)
    if (approvalEvents.length === 0) {
      return []
    }

    for (const event of approvalEvents) {
      if (event.args.owner === this.csAccountingAddress) {
        const f: Finding = new Finding()
        f.setName(`ðŸ”µ Lido stETH: Approval`)
        f.setDescription(`${event.args.spender} received allowance from ${event.args.owner} to ${event.args.value}`)
        f.setAlertid('STETH-APPROVAL')
        f.setSeverity(Finding.Severity.INFO)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)
      }
    }
    return out
  }

  public handleTransaction(txEvent: TransactionDto): Finding[] {
    const out: Finding[] = []

    const csAccountingFindings = handleEventsOfNotice(txEvent, this.csAccountingEvents)
    const stETHApprovalFindings = this.handleStETHApprovalEvents(txEvent)

    out.push(...csAccountingFindings, ...stETHApprovalFindings)

    return out
  }
}
