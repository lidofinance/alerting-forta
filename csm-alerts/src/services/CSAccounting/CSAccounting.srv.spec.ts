import { DeploymentAddress, DeploymentAddresses } from '../../utils/constants.holesky'
import { expect } from '@jest/globals'
import { TransactionDto } from '../../entity/events'
import {
  CSModule__factory,
  CSAccounting__factory,
  CSFeeDistributor__factory,
  CSFeeOracle__factory,
} from '../../generated/typechain'
import { getCSAccountingEvents } from '../../utils/events/cs_accounting_events'
import { CSAccountingSrv, ICSAccountingClient } from './CSAccounting.srv'
import * as Winston from 'winston'
import { ETHProvider } from '../../clients/eth_provider'
import { ethers } from 'forta-agent'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'
import promClient from 'prom-client'
import { Metrics } from '../../utils/metrics/metrics'

const TEST_TIMEOUT = 120_000 // ms

describe('CSAccounting event tests', () => {
  const chainId = 17000

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const address: DeploymentAddress = DeploymentAddresses

  const fortaEthersProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)
  const csModuleRunner = CSModule__factory.connect(address.CS_MODULE_ADDRESS, fortaEthersProvider)
  const csAccountingRunner = CSAccounting__factory.connect(address.CS_ACCOUNTING_ADDRESS, fortaEthersProvider)
  const csFeeDistributorRunner = CSFeeDistributor__factory.connect(
    address.CS_FEE_DISTRIBUTOR_ADDRESS,
    fortaEthersProvider,
  )
  const csFeeOracleRunner = CSFeeOracle__factory.connect(address.CS_FEE_ORACLE_ADDRESS, fortaEthersProvider)

  const registry = new promClient.Registry()
  const m = new Metrics(registry, 'test_')

  const csAccountingClient: ICSAccountingClient = new ETHProvider(
    logger,
    m,
    fortaEthersProvider,
    csModuleRunner,
    csAccountingRunner,
    csFeeDistributorRunner,
    csFeeOracleRunner,
  )

  const csAccountingSrv = new CSAccountingSrv(
    logger,
    csAccountingClient,
    getCSAccountingEvents(address.CS_ACCOUNTING_ADDRESS),
    address.CS_ACCOUNTING_ADDRESS,
    address.LIDO_STETH_ADDRESS,
    address.CS_MODULE_ADDRESS,
  )

  test(
    'ðŸš¨ CSAccounting: Bond Curve Updated',
    async () => {
      const txHash = '0x8b904da83d58e520c778cc562b10fa4e0943a9f991b9050d93481fdabf2da9c2'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csAccountingSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ CSAccounting: Bond Curve added, stETH Approval',
    async () => {
      const txHash = '0xcc92653babec3b1748d8e04de777796cab2d1ae40fbe926db857e9103a9b74a5'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csAccountingSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(4)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸš¨ CSAccounting: Charge Penalty Recipient Set',
    async () => {
      const txHash = '0xf62269919009e1cb9c6ea8c29cb6c83f9c1d113d97d401ed5ff2b696cee6d82f'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csAccountingSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ CSAccounting: Bond Curve Set',
    async () => {
      const txHash = '0xcea4d214c8f6e4f3415fc941fdb6802f4243a7b3e12ba5288cf7e7df39d457a0'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csAccountingSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )
})
