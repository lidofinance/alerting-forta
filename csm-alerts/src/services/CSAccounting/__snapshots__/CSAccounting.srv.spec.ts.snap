// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSAccounting event tests ðŸ”´ CSAccounting: Bond Curve Set 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CS-ACCOUNTING-BOND-CURVE-SET",
      "ðŸ”´ CSAccounting: Bond curve set",
      "Bond curve set for node operator ID 204 with curve ID 3",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "204,3",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSAccounting event tests ðŸ”´ CSAccounting: Bond Curve added, stETH Approval 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CS-ACCOUNTING-BOND-CURVE-ADDED",
      "ðŸ”´ CSAccounting: Bond curve added",
      "Bond curve added: [2.000 ETH, 3.900 ETH, 5.700 ETH, 7.400 ETH, 9.000 ETH, 10.500 ETH]",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "2000000000000000000,3900000000000000000,5700000000000000000,7400000000000000000,9000000000000000000,10500000000000000000",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      1,
      ,
      4,
      "STETH-APPROVAL",
      "ðŸ”µ Lido stETH: Approval",
      "0x8d09a4502Cc8Cf1547aD300E066060D043f6982D received allowance from 0xc093e53e8F4b55A223c18A2Da6fA00e60DD5EFE1 to 115792089237316195423570985008687907853269984665640564039457584007913129639935",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": null,
  },
  {
    "array": [
      "ethereum",
      1,
      ,
      4,
      "STETH-APPROVAL",
      "ðŸ”µ Lido stETH: Approval",
      "0xc7cc160b58F8Bb0baC94b80847E2CF2800565C50 received allowance from 0xc093e53e8F4b55A223c18A2Da6fA00e60DD5EFE1 to 115792089237316195423570985008687907853269984665640564039457584007913129639935",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": null,
  },
  {
    "array": [
      "ethereum",
      1,
      ,
      4,
      "STETH-APPROVAL",
      "ðŸ”µ Lido stETH: Approval",
      "0x4E46BD7147ccf666E1d73A3A456fC7a68de82eCA received allowance from 0xc093e53e8F4b55A223c18A2Da6fA00e60DD5EFE1 to 115792089237316195423570985008687907853269984665640564039457584007913129639935",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": null,
  },
]
`;

exports[`CSAccounting event tests ðŸš¨ CSAccounting: Bond Curve Updated 1`] = `
[
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "CS-ACCOUNTING-BOND-CURVE-UPDATED",
      "ðŸš¨ CSAccounting: Bond curve updated",
      "Bond curve updated with curve ID 3. Bond curve: [4.000 ETH, 5.000 ETH, 6.000 ETH, 7.000 ETH]",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "3,4000000000000000000,5000000000000000000,6000000000000000000,7000000000000000000",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSAccounting event tests ðŸš¨ CSAccounting: Charge Penalty Recipient Set 1`] = `
[
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "CS-ACCOUNTING-CHARGE-PENALTY-RECIPIENT-SET",
      "ðŸš¨ CSAccounting: Charge penalty recipient set",
      "Charge penalty recipient set to [0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d](https://etherscan.io/address/0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d) (expecting the treasury)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;
